### ISSUE TITLE
Elastic APM HTTP error (400): Problem validating JSON document against schema: unexpected EOF

### ISSUE THREAD

**Issue by Sonosar** [elastic/apm-agent-nodejs#141]
When we have invalid requests with no JSON body at all APM agent crashes with this output to console
All request with contentType : application/json should actually send at lease "{}", but anyways I think APM agent should handle this properly.
My only guess that it is not handling this properly is actually error log, if it was an info level log I'd consider it safe.

**Comment by watson** [comment]
The issue is that the APM Server didn't get the entire body of JSON sent from the agent. Not necessarily that the body was empty.

The error message it self is expected in such a case, but the agent shouldn't crash when outputting it. Do you mean that the process dies, or just that it outputs the error to STDERR?

It would be really interesting to know what the agent tried to send to the APM Server that couldn't be parsed. Are you able to reproduce this problem?

**Comment by watson** [comment]
I just realized this error could be happening because the payload sent to the APM Server is too large. If your app is generating a lot of transactions or if some of the transactions are very large, the resulting payload might trigger this error.

Here's a few ways to work around that:

- The current default is that the recorded transactions are sent to the APM Server every 60 seconds (we'll lower this default to somewhere between 10 and 30 seconds soon). But you can overwrite it already now using the [`flushInterval`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/agent-api.html#flush-interval) config option
- You can also trigger a flush of the queue when it reaches a certain number of transactions using the [`maxQueueSize`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/agent-api.html#max-queue-size) config option
- Another config option that can influence the size of the payload is the [`stackTraceLimit`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/agent-api.html#stack-trace-limit) option. If your application ends up in an infinite loop that generates a stack overflow, it might be worth modifying this option to limit the number of stack frames included in the payload (we'll probably change the default value for this config option soon to a lower value, e.g. `100`)

You can try to tune one of these 3 config options to lower the size of the payload to get rid of the error.

Try for instance these settings:

```js
{
  flushInterval: 10,
  maxQueueSize: 100,
  stackTraceLimit: 100
}
```

Let me know if it works 😃

**Comment by watson** [comment]
The APM Server will by default accept JSON documents of up to 10MB (unzipped). You can also modify this limit by starting the APM Server with `-E apm-server.max_unzipped_size=<bytes>`

**Comment by Sonosar** [comment]
Couple of notes to take into action
1.Our apps are sending messages with empty JSON messages but with content type set to 
    application/json.
2.I deployed this to production, so we had high volume of transaction going in.
3.We have big messages.
4 not sure if it matters but we are utilizing websockets with socket.io 

I'll setup separate server and will send only small portion of production traffic.

On a side note, I'm very excited with this product and will track its development.

**Comment by watson** [comment]
I don't think the issue is related to your apps ending messages with empty JSON messages - unless it's sending these messages to the APM Server?

I'm curious how much traffic your production app is receiving (e.g. requests per minute) as I think the issue is related to that. Of course we should be able to handle your load no matter what, so I want to make sure our default values are set up correctly if that indeed is the issue.

You also mentioned that your app crashed due to this error? Was that correctly understood? I ask because that seems very odd. It should simply print it to the error log, but it should not crash, so I'd like to investigate this if that's the case.

**Comment by Sonosar** [comment]
It is about 3000 RPM
This does not count the traffic from websocket though. (websockets take ~80% of our traffic)
Some of the responses will have about 60kb data returned.
Regarding the crash:
    We did not experience crash actually , pm2 didn't restart the process , for example.
    What I did experience was inconsistent behavior after this messages.
    BTW I was getting 
**Error: write ECONNRESET
     at exports._errnoException (util.js:1020:11)
     at WriteWrap.afterWrite (net.js:800:14)**
I thought this 2 errors might be related , most likely this error message is the reason.

**Comment by watson** [comment]
Thanks for the details. We'll investigate further 👍

**Comment by mikker** [comment]
Hey @fracklen! Nice catch. Some kind of a validation would be a nice addition. FYI you don't have to set the two `framework_` settings as the APM Agent finds them itself. You can override them though if this is what you want :smile

**Comment by watson** [comment]
For reference, this^ is related to the [Ruby agent](https://github.com/elastic/apm-agent-ruby), not the Node agent 🤓

**Comment by phani8996** [comment]
Facing the same issue with node-agent.

2018/03/01 08:33:29.986483 handlers.go:226: ERR Problem validating JSON document against schema: I[#] S[#] doesn't validate with "transaction#"
  I[#] S[#/required] missing properties: "app", code=400

apm-server version 6.0.0-rc2 (amd64), libbeat 6.0.0-rc2
Apm-client node version: elastic-apm-node@1.1.0

**Comment by mikker** [comment]
@phani8996 `app` was renamed to `service`. You should upgrade your apm-server version to 6.2

**Comment by watson** [comment]
@phani8996 The Node.js agent has a guide for upgrading from 0.x to 1.x: https://www.elastic.co/guide/en/apm/agent/nodejs/master/upgrade-to-v1.html

**Comment by watson** [comment]
@Sonosar It has taken us some time to reproduce this problem, but I think we now know what's going on. We have identified that the APM Server in some cases can grind to a halt while processing requests if it's under very heavy load. We're working on fixing this and in general reducing its resource requirements.

To work around this in the meantime, and to ensure that your APM Server is scaled correctly in general, we recommend that you deploy multiple APM Servers behind a load balancer (or install one APM Server on each machine that are running your app - as long as each individual app process doesn't generate too much traffic).

For more info about high availability, read: https://www.elastic.co/guide/en/apm/server/6.2/high-availability.html

**Comment by alvarolobato** [comment]
We expect this to be fixed in 6.3

**Comment by watson** [comment]
This should be fixed with APM Server v6.3, so I'll close this issue for now. If it's still a problem, don't hesitate to reopen 😃
